src/toil/worker.py

```diff
@@ -259,10 +259,6 @@ def badWorker():
                 else:
                     jobGraph.command = jobGraph.checkpoint
 
-                # Reduce the retry count
-                assert jobGraph.remainingRetryCount >= 0
-                jobGraph.remainingRetryCount = max(0, jobGraph.remainingRetryCount - 1)
```

Comment:
The remainingRetryCount never got below 0 (at least I don't think so). The assert was removed when I moved the retry-count decrementing to the leader. I could have left the assert in, but didn't feel that there was much point now that the leader handles things.

When a checkpoint fails, the "cleanup" of its successors is only performed on the next run of the checkpoint. That's also the case on resume: the cleanup of the checkpoint is only ever handled by the checkpoint.

So if you have this situation, say with a retryCount of 1 for the workflow:

- workflow runs, CheckpointJob runs and creates a bunch of ChildJobs
- one ChildJob totally fails and kills the workflow
- on resume, CheckpointJob is run and removes all its successors. It reduces its remainingRetryCount to 0. It then runs from a fresh start and creates a bunch of ChildJobs.
- All its ChildJobs succeed. The workflow finishes.
- The workflow is considered totally failed because a totally failed job is detected (CheckpointJob).
- On restart you start from CheckpointJob yet again, wasting all the (successful) work done earlier. If it succeeds this time, everything is OK.

The behavior is similar with a higher retryCount, but to trigger this bug the checkpoint has to be restarted retryCount - 1 times.

I totally agree that this PR causes more blocking I/O on the leader. We could instead call this code to "cleanup" the checkpoint as part of the general jobStore "cleanup" when resuming, which would avoid leader I/O. But this was a huge blocker for me, so just needed to fix it ASAP at the time.